File: lab-redes/python/multicast/cliente_multicast.py
import os
import socket
import struct
import sys

GRUPO_MULTICAST = "230.0.0.1"
PORTA_BASE = 4446
INTERFACE_LOCAL = "127.0.0.1"


def ler_offset() -> int:
    if len(sys.argv) > 1 and sys.argv[1].isdigit():
        return int(sys.argv[1])
    return int(os.getenv("LAB_REDES_OFFSET", "0"))


def ler_max_mensagens() -> int:
    if len(sys.argv) > 2 and sys.argv[2].isdigit():
        return int(sys.argv[2])
    return int(os.getenv("LAB_REDES_MAX_MENSAGENS", "0"))


porta = PORTA_BASE + ler_offset()
max_mensagens = ler_max_mensagens()

with socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP) as sock:
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind(("", porta))

    grupo = socket.inet_aton(GRUPO_MULTICAST)
    interface = socket.inet_aton(INTERFACE_LOCAL)
    solicitacao_membro = struct.pack("4s4s", grupo, interface)
    sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, solicitacao_membro)

    print(f"[Multicast] Inscrito no grupo {GRUPO_MULTICAST}:{porta}.", flush=True)
    recebidas = 0

    while max_mensagens == 0 or recebidas < max_mensagens:
        dados, _ = sock.recvfrom(1024)
        recebidas += 1
        print(f"[Multicast] Recebido: {dados.decode('utf-8')}", flush=True)

File: lab-redes/java/multicast/ClienteMulticast.java
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.MulticastSocket;
import java.net.NetworkInterface;

public class ClienteMulticast {
    private static final int PORTA_BASE = 4446;
    private static final String GRUPO_MULTICAST = "230.0.0.1";

    public static void main(String[] args) throws Exception {
        int offset = lerOffset(args);
        int porta = PORTA_BASE + offset;
        int maxMensagens = lerMaxMensagens(args);

        InetAddress grupo = InetAddress.getByName(GRUPO_MULTICAST);
        InetSocketAddress endpointGrupo = new InetSocketAddress(grupo, porta);
        NetworkInterface loopback = NetworkInterface.getByInetAddress(InetAddress.getLoopbackAddress());

        try (MulticastSocket socket = new MulticastSocket(null)) {
            socket.setReuseAddress(true);
            socket.bind(new InetSocketAddress(porta));
            socket.joinGroup(endpointGrupo, loopback);

            System.out.println("[Multicast] Inscrito no grupo " + GRUPO_MULTICAST + ":" + porta + ".");
            byte[] buffer = new byte[1024];
            int recebidas = 0;

            while (maxMensagens == 0 || recebidas < maxMensagens) {
                DatagramPacket pacote = new DatagramPacket(buffer, buffer.length);
                socket.receive(pacote);
                String mensagem = new String(pacote.getData(), 0, pacote.getLength());
                recebidas++;
                System.out.println("[Multicast] Recebido: " + mensagem);
            }

            socket.leaveGroup(endpointGrupo, loopback);
        }
    }

    private static int lerOffset(String[] args) {
        if (args.length > 0 && args[0].matches("\\d+")) {
            return Integer.parseInt(args[0]);
        }

        String valor = System.getenv("LAB_REDES_OFFSET");
        return valor == null || valor.isBlank() ? 0 : Integer.parseInt(valor);
    }

    private static int lerMaxMensagens(String[] args) {
        if (args.length > 1 && args[1].matches("\\d+")) {
            return Integer.parseInt(args[1]);
        }

        String valor = System.getenv("LAB_REDES_MAX_MENSAGENS");
        return valor == null || valor.isBlank() ? 0 : Integer.parseInt(valor);
    }
}
